packages/material-ui/src/IconButton/IconButton.js

```diff
@@ -27,76 +67,79 @@ export const styles = (theme) => ({
         backgroundColor: 'transparent',
       },
     },
-    '&$disabled': {
-      backgroundColor: 'transparent',
-      color: theme.palette.action.disabled,
-    },
-  },
-  /* Styles applied to the root element if `edge="start"`. */
-  edgeStart: {
-    marginLeft: -12,
-    '$sizeSmall&': {
-      marginLeft: -3,
-    },
-  },
-  /* Styles applied to the root element if `edge="end"`. */
-  edgeEnd: {
-    marginRight: -12,
-    '$sizeSmall&': {
-      marginRight: -3,
-    },
-  },
-  /* Styles applied to the root element if `color="inherit"`. */
-  colorInherit: {
-    color: 'inherit',
-  },
-  /* Styles applied to the root element if `color="primary"`. */
-  colorPrimary: {
-    color: theme.palette.primary.main,
-    '&:hover': {
-      backgroundColor: alpha(theme.palette.primary.main, theme.palette.action.hoverOpacity),
-      // Reset on touch devices, it doesn't add specificity
-      '@media (hover: none)': {
-        backgroundColor: 'transparent',
+    /* Styles applied to the root element if `edge="start"`. */
+    ...(styleProps.edge === 'start' && {
+      marginLeft: styleProps.size === 'small' ? -3 : -12,
+    }),
+    /* Styles applied to the root element if `edge="end"`. */
+    ...(styleProps.edge === 'end' && {
+      marginRight: styleProps.size === 'small' ? -3 : -12,
+    }),
+    /* Styles applied to the root element if `color="inherit"`. */
+    ...(styleProps.color === 'inherit' && {
+      color: 'inherit',
+    }),
+    /* Styles applied to the root element if `color="primary"`. */
+    ...(styleProps.color === 'primary' && {
```

Comment:
Let's make each of the consecutive conditional styles callback functions, as otherwise we would completely replace the initial `'&:hover'` styles.

Reply:
Or at least where we are defining hover styles.

Reply:
How's this looking?

Reply:
Good 👍